gensim/models/keyedvectors.py

```diff
@@ -559,6 +560,90 @@ def similar_by_vector(self, vector, topn=10, restrict_vocab=None):
+                          nonzero_limit=100, dtype=REAL):
+        """Constructs a term similarity matrix for computing Soft Cosine Similarity.
+
+        Constructs a a sparse term similarity matrix in the `scipy.sparse.csc_matrix` format for
+        computing Soft Cosine Similarity between documents.
+
+        Parameters
+        ----------
+        corpus : list of lists of (int, float) two-tuples
+            A list of documents in the gensim document format.
+        dictionary : gensim.corpora.Dictionary
+            A dictionary associated with the corpus.
+        threshold : float, optional
+            Only pairs of words whose embeddings are more similar than `threshold` are considered
+            when building the sparse term similarity matrix. Defaults to `0.0`.
+        exponent : float, optional
+            The exponent applied to the similarity between two word embeddings when building the
+            term similarity matrix. Defaults to `2.0`.
+        nonzero_limit : int, optional
+            The maximum number of non-zero elements outside the diagonal in a single row or column
+            of the term similarity matrix. Setting `nonzero_limit` to a constant ensures that the
+            time complexity of computing the Soft Cosine Similarity will be linear in the document
+            length rather than quadratic. Defaults to `100`.
+        dtype : numpy.dtype, optional
+            Data-type of the term similarity matrix. Defaults to `numpy.float32`.
+
+        Returns
+        -------
+        scipy.sparse.csc_matrix
+            The constructed term similarity matrix.
+
+        See Also
+        --------
+        gensim.matutils.softcossim
+            The Soft Cosine Similarity.
+        gensim.similarities.docsim.SoftCosineSimilarity
+            A class for performing corpus-based similarity queries with Soft Cosine Similarity.
+
+        References
+        ----------
+        The constructed matrix corresponds to the matrix Mrel defined in section 2.1 of
+        [charletdamnati17]_.
+
+        .. [charletdamnati17] Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017
+           Task 3: Soft-Cosine Semantic Similarity between Questions for Community Question
+           Answering", 2017.
+        """
+
+        logger.info("constructing a term similarity matrix")
+        similarity_matrix = sparse.identity(len(dictionary), dtype=dtype, format="lil")
+        # Traverse rows.
+        num_rows = len(dictionary)
+        for w1_index in range(num_rows):
+            if w1_index % ceil(num_rows / 10) == 0:
+                logger.info("PROGRESS: at row %i / %i", w1_index + 1, num_rows)
+            w1 = dictionary[w1_index]
+            if w1 not in self.vocab:
+                continue  # A word from the dictionary not present in the word2vec model.
+            # Traverse upper triangle columns.
+            if len(dictionary) <= nonzero_limit + 1:  # Traverse all columns.
```

Comment:
`if num_rows` instead of `len(dictionary)` ?

Reply:
Fixed in effef71902ff9f0d67b79467c8411bbeae9c9445.